### PR TITLE
feat: [PLG-2024]: Credits provisioning 

### PR DESCRIPTION
**Pull request by Xin-Harness**
## Describe your changes
Use dummy hardcode value to test the end to end workflow of provisioning credits.
image
Please note that the above data does not includes all the properties of credit, it is just a dummy test to test the end to end work flow and see if the entry in credits table can be preserved. 
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)

### CODE REVIEW

File: 970-ng-commons/src/main/java/io/harness/CreditType.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  enum CreditVisibility { INTERNAL, PUBLIC }
+
+  CreditType(String displayName, CreditVisibility visibility) {
+    this.displayName = displayName;
+    this.visibility = visibility;
+  }
+
+  @JsonCreator(mode = JsonCreator.Mode.DELEGATING)
+  public static CreditType fromString(String creditType) {
+    for (CreditType creditEnum : CreditType.values()) {
```
Comment (thomas-grasman) [inline 970-ng-commons/src/main/java/io/harness/CreditType.java:34]: You can just use 
`    return CreditType.fromString(creditType);`

File: 930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+@Hidden
+@OwnedBy(HarnessTeam.PLG)
+public class CreditResource {
+  private CreditService creditService;
+  /**
+   * Create the credit. Todo: discuss with frontend about the payload etc.
+   */
+  @GET
+  @Path("/purchase")
+  @PublicApi
```
Comment (thomas-grasman) [inline 930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java:50]: What happens when the resource is set to @Hidden, but the endpoint is @PublicApi?
Are you sure we should set purchasing credits to public?

**Review by neelam-harness — CHANGES_REQUESTED** [review]
> Reviewed 6 of 11 files at r1, 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 11 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRz1Fu0SdmURS9Pzmv:-NQRz1Fu0SdmURS9Pzmw:bswadvi) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java#L42)):*
> ```Java
>   @NotEmpty protected CreditStatus creditStatus;
>   @NotEmpty protected int quantity;
>   @NotEmpty protected long purchaseTime;
> ```

Did we miss adding `int purchasePrice` here or is this information going to be captured somewhere else ?
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java` line 43 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRwBlS358_M_r4GAym:-NQRwBlS358_M_r4GAyn:bw279fr) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java#L43)):*
> ```Java
>   @NotEmpty protected int quantity;
>   @NotEmpty protected long purchaseTime;
>   @NotEmpty protected long expiryTime;
> ```

should we add a note here saying,  "creditStatus needs to be updated to EXPIRED when the expiryTime < currentTime? " or is there a better way to have this enforcement done?
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 35 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRxUES8EoVWNn7HvUu:-NQRxUES8EoVWNn7HvUv:bnfkpda) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L35)):*
> ```Java
> 
> @ApiResponses(value =
>     {
> ```

do we also need to add  `@ApiResponse(code = 200` here is it by default added ?



_Code quote:_
```Java
 @ApiResponse(code = 400, response = FailureDTO.class
```
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java` line 29 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRxrTg0W3SqZtPqSzc:-NQRxrTg0W3SqZtPqSzd:b-ce0sxr) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java#L29)):*
> ```Java
>   public void purchaseCredits(String accountIdentifier) {
>     // Todo: need to determine the details later on, so fat its a temporary dummy code
>     Credit buildCredits = CICredit.builder().build();
> ```

nit: can we just name it as `ciCredit` instead of `buildCredits` ?

**Review by Xin-Harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 11 file(s) to review. 1 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRz1Fu0SdmURS9Pzmv:-NQVWvEDEEZGPRWDJD6_:bnmhg4q) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java#L42)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Did we miss adding `int purchasePrice` here or is this information going to be captured somewhere else ?
</blockquote></details>

Thats a very good point, the purchase price is not communicated with PM, how we would want to provision the credit at what price etc. Will add it in the future and not in this PR.
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java` line 43 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRwBlS358_M_r4GAym:-NQSpS7s1rqfdZmzr_8g:b1ohjz) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java#L43)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

should we add a note here saying,  "creditStatus needs to be updated to EXPIRED when the expiryTime < currentTime? " or is there a better way to have this enforcement done?
</blockquote></details>

This logic will handle later on by the expiry handler, since status change to EXPIRED is obvious when expiryTime< currentTime, therefore won't add the comment for code convention reason.
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 50 at r1](https://reviewable.io/reviews/harness/harness-core/44992#-NQQezkC8Yj52qUoTOSt-r1-50:-NQSpGZc2ChGHVo7ZS-o:bs7wumx) ([raw file](https://github.com/harness/harness-core/blob/be08fdc5e37c4badfc5c1df92904e56447ef48b7/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L50)):*
<details><summary><i>Previously, thomas-grasman (Thomas Grasman) wrote…</i></summary><blockquote>

What happens when the resource is set to @Hidden, but the endpoint is @PublicApi?
Are you sure we should set purchasing credits to public?
</blockquote></details>

Done. 
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 35 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRxUES8EoVWNn7HvUu:-NQSpNf6DJpjVA5k1F6i:b-3gypb8) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L35)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

do we also need to add  `@ApiResponse(code = 200` here is it by default added ?
</blockquote></details>

By default, it is 200. No need to add it.
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java` line 29 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRxrTg0W3SqZtPqSzc:-NQSohuz072wrUpxG0UT:bjzeqdr) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java#L29)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

nit: can we just name it as `ciCredit` instead of `buildCredits` ?
</blockquote></details>

This is a dummy code as comment said, so it won't take any affect. Will change the entire code once the UX design is fixed.
___
*[`970-ng-commons/src/main/java/io/harness/CreditType.java` line 34 at r1](https://reviewable.io/reviews/harness/harness-core/44992#-NQBSeuq_xBHdzye9Dno-r1-34:-NQSpEgjCBTc06qBi1bZ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/be08fdc5e37c4badfc5c1df92904e56447ef48b7/970-ng-commons/src/main/java/io/harness/CreditType.java#L34)):*
<details><summary><i>Previously, thomas-grasman (Thomas Grasman) wrote…</i></summary><blockquote>

You can just use 
`    return CreditType.valueOf(creditType);`
</blockquote></details>

Done.

**Review by thomas-grasman — APPROVED** [review]
> Reviewed 8 of 11 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 3 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained.

**Review by Xin-Harness — COMMENTED** [review]
> Reviewed 8 of 11 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 3 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained.

**Review by neelam-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 11 files at r1, 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 1 change requests. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRz1Fu0SdmURS9Pzmv:-NQVudyB63C-f2xMkJSS:b93su3y) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java#L42)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

Thats a very good point, the purchase price is not communicated with PM, how we would want to provision the credit at what price etc. Will add it in the future and not in this PR.
</blockquote></details>

Sure, just wanted to see if didn't miss it. Sounds good.
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java` line 43 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRwBlS358_M_r4GAym:-NQVu__p1Gsw6YetzH_K:bgmzzzj) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java#L43)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

This logic will handle later on by the expiry handler, since status change to EXPIRED is obvious when expiryTime< currentTime, therefore won't add the comment for code convention reason.
</blockquote></details>

Sounds good. Thanks
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 49 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQVye2tBB3u_TRzg_VL:-NQVye2tBB3u_TRzg_VM:bzc66uf) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L49)):*
> ```Java
>   @POST
>   @Path("/purchase")
>   public RestResponse<Void> purchaseCredits() {
> ```

As discussed in this morning's sync, I think it'll be better to rename it as `createCredit` vs `purchaseCredits` since the purchase happens somewhere else and we are only provisioning it. 
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java` line 29 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRxrTg0W3SqZtPqSzc:-NQVuRN9AbvYLAC57rJl:bgmzzzj) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java#L29)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

This is a dummy code as comment said, so it won't take any affect. Will change the entire code once the UX design is fixed.
</blockquote></details>

Sounds good. Thanks

**Review by neelam-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 1 change requests. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 49 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQVye2tBB3u_TRzg_VL:-NQVzDtv5miUOsc-Ig7P:byy2xbw) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L49)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

As discussed in this morning's sync, I think it'll be better to rename it as `createCredit` vs `purchaseCredits` since the purchase happens somewhere else and we are only provisioning it. 
</blockquote></details>

One more thing, also is this API going to be a bulk create API or an single credit creation API at a time ? 
If it's going to `createCredits` then we will have to take a list in the API signature, I think.

**Review by Xin-Harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQRz1Fu0SdmURS9Pzmv:-NQWQea67zhK9wnn_-hv:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/entities/Credit.java#L42)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Sure, just wanted to see if didn't miss it. Sounds good.
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 49 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQVye2tBB3u_TRzg_VL:-NQWQk8T0P3LeM-J-6LS:b-8l3ox5) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L49)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

One more thing, also is this API going to be a bulk create API or an single credit creation API at a time ? 
If it's going to `createCredits` then we will have to take a list in the API signature, I think.
</blockquote></details>

Hi @neelam-harness , I will make the API change for the next PR since this API is just a stake holder for now. I will rename it from purchaseCredits to createCredit. I will consider the request and response type in the signature as well.

**Review by Xin-Harness — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 2 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained.

**Review by Xin-Harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 2 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 49 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQVye2tBB3u_TRzg_VL:-NQWhXQs64NsLD_phPfD:bpii59a) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L49)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

Hi @neelam-harness , I will make the API change for the next PR since this API is just a stake holder for now. I will rename it from purchaseCredits to createCredit. I will consider the request and response type in the signature as well.
</blockquote></details>

Renamed in this PR, done

**Review by neelam-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 49 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQVye2tBB3u_TRzg_VL:-NQWhwdDBr1vOhfDHFrW:b-hay2ib) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L49)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

Renamed in this PR, done
</blockquote></details>

Thanks!

**Review by thomas-grasman — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 1 file(s) to review.

**Review by thomas-grasman — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 1 file(s) to review.

**Review by Xin-Harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 1 file(s) to review. 1. unresolved discussions. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java` line 49 at r2](https://reviewable.io/reviews/harness/harness-core/44992#-NQVye2tBB3u_TRzg_VL:-NQWqwPC0RCu2CX2MnZN:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2f438e02c7b96c556ccf31361586e70722968c77/930-ng-license-manager/src/main/java/io/harness/credit/resource/CreditResource.java#L49)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Thanks! 
</blockquote></details>

Done.

**Review by thomas-grasman — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: 1 file(s) to review. 1. unresolved discussions.

**Review by neelam-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44992)* status: :shipit: complete! ✓

**Pull request merged**